File: Editorial/64_Minimum_Path_Sum/TypeScript/3_Dynamic_Programming_1D.ts
function minPathSum(grid: number[][]): number {
    let dp: number[] = new Array(grid[0].length).fill(0);
    for (let i = grid.length - 1; i >= 0; i--) {
        for (let j = grid[0].length - 1; j >= 0; j--) {
            if (i == grid.length - 1 && j != grid[0].length - 1)
                dp[j] = grid[i][j] + dp[j + 1];
            else if (j == grid[0].length - 1 && i != grid.length - 1)
                dp[j] = grid[i][j] + dp[j];
            else if (i != grid.length - 1 && j != grid[0].length - 1)
                dp[j] = grid[i][j] + Math.min(dp[j], dp[j + 1]);
            else dp[j] = grid[i][j];
        }
    }
    return dp[0];
}

File: Editorial/0064_Minimum_Path_Sum/Go/3_Dynamic_Programming_1D.go
func minPathSum(grid [][]int) int {
    dp := make([]int, len(grid[0]))
    for i := len(grid) - 1; i >= 0; i-- {
        for j := len(grid[0]) - 1; j >= 0; j-- {
            if i == len(grid)-1 && j != len(grid[0])-1 {
                dp[j] = grid[i][j] + dp[j+1]
            } else if j == len(grid[0])-1 && i != len(grid)-1 {
                dp[j] = grid[i][j] + dp[j]
            } else if i != len(grid)-1 && j != len(grid[0])-1 {
                dp[j] = grid[i][j] + min(dp[j], dp[j+1])
            } else {
                dp[j] = grid[i][j]
            }
        }
    }
    return dp[0]
}

func min(a int, b int) int {
    if a < b {
        return a
    }
    return b
}
